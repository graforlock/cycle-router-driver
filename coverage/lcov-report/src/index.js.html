<!doctype html>
<html lang="en">
<head>
    <title>Code coverage report for src/index.js</title>
    <meta charset="utf-8" />
    <link rel="stylesheet" href="../prettify.css" />
    <link rel="stylesheet" href="../base.css" />
    <meta name="viewport" content="width=device-width, initial-scale=1">
    <style type='text/css'>
        .coverage-summary .sorter {
            background-image: url(../sort-arrow-sprite.png);
        }
    </style>
</head>
<body>
<div class='wrapper'>
  <div class='pad1'>
    <h1>
      <a href="../index.html">All files</a> / <a href="index.html">src</a> index.js
    </h1>
    <div class='clearfix'>
      <div class='fl pad1y space-right2'>
        <span class="strong">78.95% </span>
        <span class="quiet">Statements</span>
        <span class='fraction'>30/38</span>
      </div>
      <div class='fl pad1y space-right2'>
        <span class="strong">55.56% </span>
        <span class="quiet">Branches</span>
        <span class='fraction'>5/9</span>
      </div>
      <div class='fl pad1y space-right2'>
        <span class="strong">57.14% </span>
        <span class="quiet">Functions</span>
        <span class='fraction'>12/21</span>
      </div>
      <div class='fl pad1y space-right2'>
        <span class="strong">78.38% </span>
        <span class="quiet">Lines</span>
        <span class='fraction'>29/37</span>
      </div>
    </div>
  </div>
  <div class='status-line medium'></div>
<pre><table class="coverage">
<tr><td class="line-count quiet">1
2
3
4
5
6
7
8
9
10
11
12
13
14
15
16
17
18
19
20
21
22
23
24
25
26
27
28
29
30
31
32
33
34
35
36
37
38
39
40
41
42
43
44
45
46
47
48
49
50
51
52
53
54
55
56
57
58
59
60
61
62
63
64
65
66
67
68
69
70
71
72
73
74
75
76
77
78
79
80
81
82
83
84
85
86
87
88</td><td class="line-coverage quiet"><span class="cline-any cline-yes">1x</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">1x</span>
<span class="cline-any cline-yes">1x</span>
<span class="cline-any cline-yes">1x</span>
<span class="cline-any cline-yes">1x</span>
<span class="cline-any cline-yes">1x</span>
<span class="cline-any cline-yes">1x</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">1x</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">1x</span>
<span class="cline-any cline-yes">1x</span>
<span class="cline-any cline-yes">1x</span>
<span class="cline-any cline-yes">1x</span>
<span class="cline-any cline-yes">1x</span>
<span class="cline-any cline-no">&nbsp;</span>
<span class="cline-any cline-yes">1x</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">2x</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">2x</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">2x</span>
<span class="cline-any cline-yes">2x</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">1x</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-no">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">1x</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">1x</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">1x</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-no">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">1x</span>
<span class="cline-any cline-no">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-no">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-no">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-no">&nbsp;</span>
<span class="cline-any cline-no">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">1x</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">2x</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">1x</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">1x</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">1x</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">1x</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span></td><td class="text"><pre class="prettyprint lang-js">const assert = require('assert');
&nbsp;
const { adapt } = require('@cycle/run/lib/adapt');
const { createBrowserHistory, createMemoryHistory } = require('history');
const historyIntent = require('./historyIntent');
const intents = require('./intents');
const UniversalRouter = require('universal-router');
const xs = require('xstream').default;
&nbsp;
&nbsp;
let history = null, unlisten = null;
&nbsp;
module.exports = {
    go: payload =&gt; ({ type: intents.GO, payload }),
    goBack: () =&gt; ({ type: intents.GO_BACK }),
    goForward: () =&gt; ({ type: intents.GO_FORWARD }),
    push: (...payload) =&gt; ({ type: intents.PUSH, payload }),
    redirect: <span class="fstat-no" title="function not covered" >pa</span>yload =&gt; (<span class="cstat-no" title="statement not covered" >{ type: intents.REDIRECT, payload })</span>,
    replace: (...payload) =&gt; ({ type: intents.REPLACE, payload }),
&nbsp;
    intents: intents,
&nbsp;
    makeRouterDriver: function makeRouterDriver(
        routes,
        { url = false, options = {} } = {}
    ) {
        assert.ok(
            routes,
            'Cycle.js router Error: `routes` cannot be undefined.'
        );
&nbsp;
        history = typeof window !== 'undefined'
            ? createBrowserHistory(options)
            : <span class="branch-1 cbranch-no" title="branch not covered" >createMemoryHistory(options);</span>
&nbsp;
        const uRouter = new UniversalRouter(routes),
            onNextState = historyIntent(history);
&nbsp;
        function routerDriver(input$) {
            input$.addListener({
                next: onNextState,
                error: <span class="fstat-no" title="function not covered" >er</span>ror =&gt; {
<span class="cstat-no" title="statement not covered" >                    assert.fail(`Cycle.js router Error: ${error}.`);</span>
                },
                complete: <span class="fstat-no" title="function not covered" >()</span> =&gt; {}
            });
&nbsp;
            const output$ = xs.create({
                start: listener =&gt; {
                    uRouter
                        .resolve(url || <span class="branch-1 cbranch-no" title="branch not covered" >history.location.pathname)</span>
                        .then(vtree =&gt; {
                            listener.shamefullySendNext(vtree);
                        })
                        .catch(<span class="fstat-no" title="function not covered" >er</span>r =&gt; <span class="cstat-no" title="statement not covered" >listener.shamefullySendError(err))</span>;
&nbsp;
                    unlisten = history.listen(<span class="fstat-no" title="function not covered" >(l</span>ocation, action) =&gt; {
<span class="cstat-no" title="statement not covered" >                        uRouter</span>
                            .resolve(location.pathname)
                            .then(<span class="fstat-no" title="function not covered" >vt</span>ree =&gt; {
<span class="cstat-no" title="statement not covered" >                                listener.shamefullySendNext(vtree);</span>
                            })
                            .catch(<span class="fstat-no" title="function not covered" >er</span>r =&gt; <span class="cstat-no" title="statement not covered" >listener.shamefullySendError(err))</span>;
                    });
                },
                stop: <span class="fstat-no" title="function not covered" >()</span> =&gt; {
<span class="cstat-no" title="statement not covered" >                    if (unlisten) {</span>
<span class="cstat-no" title="statement not covered" >                        unlisten();</span>
                    }
                }
            });
            return adapt(output$);
        }
        return routerDriver;
    },
    mockRouterSource: function mockRouterSource(routes, { url }) {
        const output$ = xs.create({
            start: listener =&gt; {
                new UniversalRouter(routes)
                    .resolve(url)
                    .then(vtree =&gt; listener.shamefullySendNext(vtree));
            },
            stop: <span class="fstat-no" title="function not covered" >()</span> =&gt; {}
        });
        return adapt(output$);
    }
};
&nbsp;</pre></td></tr>
</table></pre>
<div class='push'></div><!-- for sticky footer -->
</div><!-- /wrapper -->
<div class='footer quiet pad2 space-top1 center small'>
  Code coverage
  generated by <a href="https://istanbul.js.org/" target="_blank">istanbul</a> at Sun Jun 25 2017 17:25:35 GMT+0100 (BST)
</div>
</div>
<script src="../prettify.js"></script>
<script>
window.onload = function () {
        if (typeof prettyPrint === 'function') {
            prettyPrint();
        }
};
</script>
<script src="../sorter.js"></script>
</body>
</html>
