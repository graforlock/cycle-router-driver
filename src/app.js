import {h1} from '@cycle/dom'
import xs from 'xstream'

export function App (sources) {
  const vtree$ = sources.Router
    .map(html => html);
  
  const click$ = sources.DOM.select('document')
    .events('click')
    .map(() =>
      h1('cliiiiiickt!') 
    );

  const merge$ = xs.merge(vtree$, click$);  
  
  const sinks = {
    DOM: merge$
  }
  return sinks
}
