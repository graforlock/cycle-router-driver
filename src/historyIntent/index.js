const intents = require('../intents');

module.exports = history => value => {
    const intentsMap = {
        [intents.GO]: payload => history.go(value.payload),
        [intents.GO_BACK]: () => history.goBack(),
        [intents.GO_FORWARD]: () => history.goForward(),
        [intents.PUSH]: (...payload) => history.push(...payload),
        [intents.REPLACE]: (...payload) => history.replace(...payload)
    };

    const handler = intentsMap[value.type];
    if (handler) {
        handler(value.payload);
        return;
    }
    history.push(value);
};
