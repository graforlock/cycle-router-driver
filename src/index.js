const assert = require('assert');

const { adapt } = require('@cycle/run/lib/adapt');
const { createBrowserHistory, createMemoryHistory } = require('history');
const historyIntent = require('./historyIntent');
const intents = require('./intents');
const UniversalRouter = require('universal-router');
const xs = require('xstream').default;


let history = null, unlisten = null;

module.exports = {
    go: payload => ({ type: intents.GO, payload }),
    goBack: () => ({ type: intents.GO_BACK }),
    goForward: () => ({ type: intents.GO_FORWARD }),
    push: (...payload) => ({ type: intents.PUSH, payload }),
    redirect: payload => ({ type: intents.REDIRECT, payload }),
    replace: (...payload) => ({ type: intents.REPLACE, payload }),

    makeRouterDriver: function makeRouterDriver(
        routes,
        { url = false, options = {} } = {}
    ) {
        assert.ok(
            routes,
            'Cycle.js router Error: `routes` cannot be undefined.'
        );

        history = typeof window !== 'undefined'
            ? createBrowserHistory(options)
            : createMemoryHistory(options);

        const uRouter = new UniversalRouter(routes),
            onNextState = historyIntent(history);

        function routerDriver(input$) {
            input$.addListener({
                next: onNextState,
                error: error => {
                    assert.fail(`Cycle.js router Error: ${error}.`);
                },
                complete: () => {}
            });

            const output$ = xs.create({
                start: listener => {
                    uRouter
                        .resolve(url || history.location.pathname)
                        .then(vtree => {
                            listener.shamefullySendNext(vtree);
                        })
                        .catch(err => listener.shamefullySendError(err));

                    unlisten = history.listen((location, action) => {
                        uRouter
                            .resolve(location.pathname)
                            .then(vtree => {
                                listener.shamefullySendNext(vtree);
                            })
                            .catch(err => listener.shamefullySendError(err));
                    });
                },
                stop: () => {
                    if (unlisten) {
                        unlisten();
                    }
                }
            });
            return adapt(output$);
        }
        return routerDriver;
    },
    mockRouterSource: function mockRouterSource(routes, { url }) {
        const output$ = xs.create({
            start: listener => {
                new UniversalRouter(routes)
                    .resolve(url)
                    .then(vtree => listener.shamefullySendNext(vtree));
            },
            stop: () => {}
        });
        return adapt(output$);
    }
};
