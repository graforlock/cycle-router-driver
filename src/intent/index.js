module.exports = {
    GO: 'GO',
    GO_BACK: 'GO_BACK',
    GO_FORWARD: 'GO_FORWARD',
    PUSH: 'PUSH',
    REPLACE: 'REPLACE'
};
