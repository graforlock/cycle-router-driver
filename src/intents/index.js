module.exports = {
    GO: 'GO',
    GO_BACK: 'GO_BACK',
    GO_FORWARD: 'GO_FORWARD',
    PUSH: 'PUSH',
    REDIRECT: 'REDIRECT',
    REPLACE: 'REPLACE'
};
