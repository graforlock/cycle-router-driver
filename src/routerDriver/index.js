import UniversalRouter from 'universal-router';
import createHistory from 'history/createBrowserHistory'
import {adapt} from '@cycle/run/lib/adapt';
import xs from 'xstream';

export function makeRouterDriver(routes) {
    const uRouter = new UniversalRouter(routes),
        history = createHistory();

    function routerDriver(input$) {
        input$.addListener({
            next: value => {
                console.log('Navigated to: ', value);
            },
            error: error => { console.error(error); },
            complete: () => {}
        });

        const incoming$ = xs.create({
            start: async listener => {
                const html = await uRouter.resolve(history.location.pathname); 
                listener.shamefullySendNext(html);
                   
                history.listen(async (location, action) => {
                    const html = await uRouter.resolve(location.pathname);
                    listener.shamefullySendNext(html);
                });   
            },
            stop: () => {}
        });
        return adapt(incoming$);
    }
    return routerDriver;
}