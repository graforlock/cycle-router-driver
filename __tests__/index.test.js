import * as uRouter from '../src/index';
import { p, makeDOMDriver } from '@cycle/dom';
import { makeHTMLDriver } from '@cycle/html';
import { run } from '@cycle/run';
import { mockTimeSource } from '@cycle/time';

import { select } from 'snabbdom-selector';
import xs from 'xstream';

describe('uRouter', () => {
    describe('unit tests', () => {
        it('exports a makeRouterDriver driver function', () => {
            expect(typeof uRouter.makeRouterDriver).toBe('function');
        });

        it('exports a makeRouterDriver driver function', () => {
            expect(uRouter.makeRouterDriver({})).toMatchSnapshot();
        });

        it('matches current API snapshot', () => {
            expect(uRouter).toMatchSnapshot();
        });

        it('go(n)', () => {
            expect(uRouter.go(+1)).toMatchSnapshot();
        });

        it('goBack()', () => {
            expect(uRouter.goBack()).toMatchSnapshot();
        });

        it('goForward()', () => {
            expect(uRouter.goForward()).toMatchSnapshot();
        });

        it('push(..any)', () => {
            expect(uRouter.push('/test')).toMatchSnapshot();
        });

        it('replace(..any)', () => {
            expect(uRouter.replace('/test')).toMatchSnapshot();
        });
    });

    describe('functional tests', () => {
        function app(sources) {
            const vtree$ = sources.Router;

            return {
                DOM: vtree$
            };
        }
        const routes = [
            {
                path: '/:param',
                action: ({ params: { param } }) => p(`Hello ${param}.`)
            }
        ];

        describe('makeRouterDriver', () => {
            it('renders a correct markup for a given route', () => {
                run(app, {
                    DOM: makeHTMLDriver(html => {
                        expect(html).toEqual('<p>Hello World.</p>');
                    }),
                    Router: uRouter.makeRouterDriver(routes, { url: '/World' })
                });
            });

            it('throws an exception on non-existent route', () => {
                try {
                    run(app, {
                        DOM: makeHTMLDriver(html => {
                            expect(html).toEqual('<p>Hello World.</p>');
                        }),
                        Router: uRouter.makeRouterDriver(routes, { url: '/bad/route' })
                    });
                } catch(err) {
                    expect(err.message).toBe('Error: Page not found');
                }
            });
        });

        describe('mockRouterSource', () => {
            it('allows users to test router source', done => {
                const Time = mockTimeSource();

                const Router = uRouter.mockRouterSource(routes, {
                    url: '/World'
                });

                const text$ = app({ Router }).DOM.map(vtree => vtree.text);

                Time.assertEqual(
                    text$,
                    Time.diagram('x', { x: 'Hello World.' })
                );
                Time.run(done);
            });
        });
    });
});
