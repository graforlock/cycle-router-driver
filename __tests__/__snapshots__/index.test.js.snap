// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`uRouter unit tests exports a makeRouterDriver driver function 1`] = `[Function]`;

exports[`uRouter unit tests go(n) 1`] = `
Object {
  "payload": 1,
  "type": "GO",
}
`;

exports[`uRouter unit tests goBack() 1`] = `
Object {
  "type": "GO_BACK",
}
`;

exports[`uRouter unit tests goForward() 1`] = `
Object {
  "type": "GO_FORWARD",
}
`;

exports[`uRouter unit tests matches current API snapshot 1`] = `
Object {
  "default": Object {
    "go": [Function],
    "goBack": [Function],
    "goForward": [Function],
    "intents": Object {
      "GO": "GO",
      "GO_BACK": "GO_BACK",
      "GO_FORWARD": "GO_FORWARD",
      "PUSH": "PUSH",
      "REDIRECT": "REDIRECT",
      "REPLACE": "REPLACE",
    },
    "makeRouterDriver": [Function],
    "mockRouterSource": [Function],
    "push": [Function],
    "redirect": [Function],
    "replace": [Function],
  },
  "go": [Function],
  "goBack": [Function],
  "goForward": [Function],
  "intents": Object {
    "GO": "GO",
    "GO_BACK": "GO_BACK",
    "GO_FORWARD": "GO_FORWARD",
    "PUSH": "PUSH",
    "REDIRECT": "REDIRECT",
    "REPLACE": "REPLACE",
  },
  "makeRouterDriver": [Function],
  "mockRouterSource": [Function],
  "push": [Function],
  "redirect": [Function],
  "replace": [Function],
}
`;

exports[`uRouter unit tests push(..any) 1`] = `
Object {
  "payload": Array [
    "/test",
  ],
  "type": "PUSH",
}
`;

exports[`uRouter unit tests replace(..any) 1`] = `
Object {
  "payload": Array [
    "/test",
  ],
  "type": "REPLACE",
}
`;
